### ISSUE TITLE
[MSHADE-291] shadedPattern applied multiples times when relocating the contents of META-INF/services files

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-shade-plugin#294]
**[Jan Luehe](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=janlu)** opened **[MSHADE-291](https://issues.apache.org/jira/browse/MSHADE-291?redirect=false)** and commented

Steps to reproduce:

1. Modified the test case for https://issues.apache.org/jira/browse/MSHADE-190, as follows:

```java
diff --git a/pom.xml b/pom.xml
index 746b700..aea9abb 100644
--- a/pom.xml
+++ b/pom.xml
@@ -68,12 +68,12 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-shade-plugin</artifactId>
-                <version>2.4</version>
+                <version>3.1.1</version>
                 <configuration>
                     <relocations>
                         <relocation>
-                            <pattern>org.eclipse.*</pattern>
-                            <shadedPattern>borg.eclipse.*</shadedPattern>
+                            <pattern>org.eclipse</pattern>
+                            <shadedPattern>org.eclipse1234</shadedPattern>
                         </relocation>
                     </relocations>
                     <transformers>
```

2. mvn package
3. jar -xvf target/shade-meta-tc-1.0-SNAPSHOT.jar META-INF/services
4. cat META-INF/services/org.osgi.framework.launch.FrameworkFactory

The shaded service implementation class looks as follows:

```java
org.eclipse12341234.osgi.launch.EquinoxFactory
```

It appears that shadedPattern was applied twice.

---

**Affects:** 3.1.1, 3.2.0, 3.2.1

**Issue Links:**
- [MSHADE-312](https://issues.apache.org/jira/browse/MSHADE-312) META-INF/services incorrectly adjusted when relocation pattern occurs twice in package name
 (_**"is duplicated by"**_)

**Remote Links:**
- [GitHub Pull Request #9
](https://github.com/apache/maven-shade-plugin/pull/9)

1 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Jan Luehe](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jluehe) even though [releasing](https://maven.apache.org/developers/release/maven-project-release-procedure.html) isn't that hard, the [voting](https://www.apache.org/foundation/voting.html) process will always claim time from several Maven PMC members. For that reason I prefer not to do releases for a single issue.
However, if you manage to provide PRs for open issues of this plugin you increase the chance of a release.

**Comment by jira-importer** [comment]
**[Lyndon Armitage](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=larmitage)** commented

What's the ETA on a release [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)? And in the meantime what is the workaround? I have spent a few days struggling with some shading issues including this one and don't see why a patch level release hasn't been made yet.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

[Lyndon Armitage](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=larmitage) even though [releasing](https://maven.apache.org/developers/release/maven-project-release-procedure.html) isn't that hard, the [voting](https://www.apache.org/foundation/voting.html) process will always claim time from several Maven PMC members. For that reason I prefer not to do releases for a single issue.
However, if you manage to provide PRs for open issues of this plugin you increase the chance of a release. (in other words, I don't have an ETA for you)

I expect using shadedPattern that doesn't contain the pattern is the workaround.

**Comment by jira-importer** [comment]
**[Jan Luehe](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jluehe)** commented

Is there any update on the 3.2.2 release?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Plans are that for Java 14 we need to update ASM. This will happen within about 1.5 months, so we'd prefer to wait for that first.
But yes, we are aware that we should do a release soon.

**Comment by jira-importer** [comment]
**[Jan Luehe](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jluehe)** commented

Confirmed the fix in the just-released 3.2.2. Thanks guys!

**Comment by jira-importer** [comment]
**[Pavel Bekkerman](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wheredevel)** commented

I have been using 3.1.1 version of the plugin, until ran into this issue.

Unfortunately, I have to report: the FIX is still NON-functional. Moreover, ServicesResourceTransformer seems to be broken completely in 3.2.2+(!)

I tried 3.2.2, 3.2.3, and 3.3.0-SNAPSHOT built from the latest `master`. 

Whatever I do, I'm getting the following error message: 

```java
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:3.3.0-SNAPSHOT:shade (default) on project XXXXXXX: Unable to parse configuration of mojo org.apache.maven.plugins:maven-shade-plugin:3.3.0-SNAPSHOT:shade for parameter resource: Cannot find 'resource' in class org.apache.maven.plugins.shade.resource.ServicesResourceTransformer -> [Help 1]
```

Here's the stack trace: 

```java
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:3.2.2:shade (default) on project XXXXXXXXXXX: Unable to parse configuration of mojo org.apache.maven.plugins:maven-shade-plugin:3.2.2:shade for parameter resource: Cannot find 'resource' in class org.apache.maven.plugins.shade.resource.ServicesResourceTransformerorg.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:3.2.2:shade (default) on project XXXXXXXXXXXX: Unable to parse configuration of mojo org.apache.maven.plugins:maven-shade-plugin:3.2.2:shade for parameter resource: Cannot find 'resource' in class org.apache.maven.plugins.shade.resource.ServicesResourceTransformer at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212) at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153) at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145) at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116) at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80) at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51) at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128) at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307) at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193) at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106) at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863) at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288) at org.apache.maven.cli.MavenCli.main(MavenCli.java:199) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:498) at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289) at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229) at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415) at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)Caused by: org.apache.maven.plugin.PluginConfigurationException: Unable to parse configuration of mojo org.apache.maven.plugins:maven-shade-plugin:3.2.2:shade for parameter resource: Cannot find 'resource' in class org.apache.maven.plugins.shade.resource.ServicesResourceTransformer at org.apache.maven.plugin.internal.DefaultMavenPluginManager.populatePluginFields(DefaultMavenPluginManager.java:662) at org.apache.maven.plugin.internal.DefaultMavenPluginManager.getConfiguredMojo(DefaultMavenPluginManager.java:594) at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:121) at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207) ... 20 moreCaused by: org.codehaus.plexus.component.configurator.ComponentConfigurationException: Cannot find 'resource' in class org.apache.maven.plugins.shade.resource.ServicesResourceTransformer at org.eclipse.sisu.plexus.CompositeBeanHelper.setProperty(CompositeBeanHelper.java:252) at org.codehaus.plexus.component.configurator.converters.composite.ObjectWithFieldsConverter.processConfiguration(ObjectWithFieldsConverter.java:101) at org.codehaus.plexus.component.configurator.converters.composite.ObjectWithFieldsConverter.fromConfiguration(ObjectWithFieldsConverter.java:57) at org.codehaus.plexus.component.configurator.converters.composite.AbstractCollectionConverter.fromChildren(AbstractCollectionConverter.java:54) at org.codehaus.plexus.component.configurator.converters.composite.ArrayConverter.fromConfiguration(ArrayConverter.java:51) at org.eclipse.sisu.plexus.CompositeBeanHelper.convertProperty(CompositeBeanHelper.java:273) at org.eclipse.sisu.plexus.CompositeBeanHelper.setProperty(CompositeBeanHelper.java:210) at org.codehaus.plexus.component.configurator.converters.composite.ObjectWithFieldsConverter.processConfiguration(ObjectWithFieldsConverter.java:101) at org.codehaus.plexus.component.configurator.BasicComponentConfigurator.configureComponent(BasicComponentConfigurator.java:34) at org.apache.maven.plugin.internal.DefaultMavenPluginManager.populatePluginFields(DefaultMavenPluginManager.java:632) ... 23 more
```

When debugging the plugin execution I was able to spot the following:

```java
...
<transformers><transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"><resource>META-INF/spring.handlers</resource>
</transformer>
</transformers>
...
```

My configuration is just:

```java
<transformers> <transformers> <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer" /> </transformers>
```

META-INF/spring.handlers is not mentioned anywhere in my project.

I know META-INF/spring.handlers is mentioned in the Transformer usage examples. I don't understand how it made its way into the execution...

Unfortunately, I wasn't able to spot its origin. But, clearly even if the FIX to the issue in context is indeed available since 3.2.2+ - We're still blocked, as ServicesResourceTransformer simply stopped working in those releases(!) 

 

In addition, I tried the following as an alternative:

```java
<transformer implementation="org.apache.maven.plugins.shade.resource.AppendingTransformer"> <transformer implementation="org.apache.maven.plugins.shade.resource.AppendingTransformer"> <resource>META-INF/services/XXXXXXXXXXXXX</resource> </transformer>
```

To my disappointment, although now the execution doesn't fail, the issue in context is NOT FIXED for AppendingTransformer: the shaded pattern is STILL applied multiples times when relocating the contents of the services file.

**Comment by jira-importer** [comment]
**[Pavel Bekkerman](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=wheredevel)** commented

1. I already figured out the reason for META-INF/spring.handlers to appear during the debug: when inheriting from spring-boot-starter-parent, some defaults are configured for maven-shade-plugin. To avoid this one can use 

```java
combine.self="override"
```

For example 

```java
<transformers combine.self="override"><transformers combine.self="override"> <transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer" /> </transformers>
```

1. As for the original issue (shaded pattern multiplication) The offending line seems to be org.apache.maven.plugins.shade.relocation.SimpleRelocator:219 

```java
return sourceContent.replaceAll( "\\b" + pattern, shadedPattern );
```

 
When pattern is null or empty (which is possible according to the documentation, and effectively allows to relocate ALL the packages) - it goes nuts.

That's the patch that worked for me:

```java
    @Override
    public String applyToSourceContent( String sourceContent )
    {
        if ( rawString )
        {
            return sourceContent;
        }
        else if ( pattern == null || pattern.isEmpty() )
        {
            return shadedPattern + sourceContent;
        }
        else
        {
            return sourceContent.replaceAll( "\\b" + pattern, shadedPattern );
        }
    }

```
